120 query parameter support

gbadner opened the pull request:
hibernate/hibernate-reactive#120

After force-pushing, this branch no longer requires https://github.com/hibernate/hibernate-orm/pull/3420.

This works for HQL queries. Some similar changes still need to be made for native queries.

More tests are needed for native queries, filters, and criteria queries.

gavinking commented:
Great! This looks excellent, @gbadner. I have a couple of comments and questions but I want to actually try it out first. Hold on.

gavinking commented:
OK, so, here's my suggestion :-)

In order to not be waiting around for a new release of Hibernate, or maintaining a development branch, I propose the following copy/paste job in `ReactiveQueryLoader`:

```java
	private ResultSet preprocessResultSet(
			ResultSet resultSet,
			final RowSelection selection,
			final LimitHandler limitHandler
	) throws SQLException, HibernateException {
		if ( !limitHandler.supportsLimitOffset()
				|| !LimitHelper.useLimit( limitHandler, selection ) ) {
			for (int i = 0,
				 firstRow = LimitHelper.getFirstRow(selection);
				 i < firstRow; i++ ) {
				resultSet.next();
			}
		}
		return resultSet;
	}

	private PreparedStatement bindPreparedStatement(
			final PreparedStatement st,
			final QueryParameters queryParameters,
			final LimitHandler limitHandler,
			final SharedSessionContractImplementor session) throws SQLException, HibernateException {

		final Dialect dialect = getFactory().getJdbcServices().getDialect();
		final RowSelection selection = queryParameters.getRowSelection();
		final boolean callable = queryParameters.isCallable();

		int col = 1;
		//TODO: can we limit stored procedures ?!
		col += limitHandler.bindLimitParametersAtStartOfQuery( selection, st, col );

		if ( callable ) {
			col = dialect.registerResultSetOutParameter( (CallableStatement) st, col );
		}

		col += bindParameterValues( st, queryParameters, col, session );

		col += limitHandler.bindLimitParametersAtEndOfQuery( selection, st, col );

		limitHandler.setMaxRows( selection, st );

		// no support for these options in Reactive
//		if ( selection != null ) {
//			if ( selection.getTimeout() != null ) {
//				st.setQueryTimeout( selection.getTimeout() );
//			}
//			if ( selection.getFetchSize() != null ) {
//				st.setFetchSize( selection.getFetchSize() );
//			}
//		}

		// handle lock timeout...
		LockOptions lockOptions = queryParameters.getLockOptions();
		if ( lockOptions != null ) {
			if ( lockOptions.getTimeOut() != LockOptions.WAIT_FOREVER ) {
				if ( !dialect.supportsLockTimeouts() ) {
					if ( LOG.isDebugEnabled() ) {
						LOG.debugf(
								"Lock timeout [%s] requested but dialect reported to not support lock timeouts",
								lockOptions.getTimeOut()
						);
					}
				}
				else if ( dialect.isLockTimeoutParameterized() ) {
					st.setInt( col++, lockOptions.getTimeOut() );
				}
			}
		}

		if ( LOG.isTraceEnabled() ) {
			LOG.tracev( "Bound [{0}] parameters total", col );
		}

		return st;
	}
```

I agree that it would be nice to make `bindPreparedStatement()` protected in the superclass, but for now this is hardly the worst copy/paste job in the whole of Hibernate Reactive, and it lets us go ahead and merge this work right away.

WDYT, @gbadner?

(If you prefer, I can push a commit to this branch.)

gavinking commented on hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/QueryParametersAdaptor.java:
```diff
@@ -7,31 +7,53 @@
-			catch (SQLException e) {
-				//can never happen
-				throw new JDBCException("error binding parameters", e);
-			}
+		try {
+			binder.bind( adaptor );
+			return adaptor.getParametersAsArray();
+		}
+		catch (SQLException e) {
+			//can never happen
+			throw new JDBCException("error binding parameters", e);
 		}
-		return adaptor.getParametersAsArray();
 	}
```
@gbadner I'm wondering if this static method would work better as a constructor of `PreparedStatementAdaptor`. WDYT?

gavinking replied:
Actually on second thoughts: I tried inlining the method and HR didn't get worse. It's not really doing much except for the exception translation. I personally would get rid of it. (This is very minor though.)

gavinking commented on hibernate-reactive-core/src/main/java/org/hibernate/reactive/stage/Stage.java:
```diff
@@ -53,6 +53,8 @@
 
 		Query<R> setParameter(int var1, Object var2);
 
+		Query<R> setParameter(String name, Object var2);
```
We also need this added to `Mutiny.Query`.

gbadner commented:
@gavinking, I'll incorporate the copy/paste to `ReactiveQueryLoader` that you suggested [here](https://github.com/hibernate/hibernate-reactive/pull/197#issuecomment-634522808).

I thought we were preferring to extending ORM over copy/paste. I understand it does complicate things until the change gets released. 

Please let me know the recommended process for things like this. 

Should I keep HHH-14044 and it's [PR](hibernate/hibernate-orm#3420) open, and try to get it incorporated into the next ORM 5.4 release, so the copy/paste code could be deleted?

gavinking commented:
> I thought we were preferring to extending ORM over copy/paste.

Well we certainly *prefer* it, of course, all I'm saying is that this is a reasonable *temporary* solution that lets us get your patch integrated right away.

> Should I keep HHH-14044 and it's PR open, and try to get it incorporated into the next ORM 5.4 release, so the copy/paste code could be deleted?

Yes, sure, though I bet if we look carefully we'll see that there's some other methods of that class that we would like to see redeclared `protected`.

gbadner commented:
OK, sounds good. Thanks @gavinking

gavinking commented:
no way thank *you*

gbadner commented:
Rebased to remove conflicts, made requested changes, reverted to use `hibernateOrmVersion = '5.4.16.Final'` and disabled  `mavenLocal()` in build.gradle.

gavinking commented:
@gbadner excellent, thanks!

Are you going to go ahead and merge it?

DavideD commented:
Reabsed and Merged https://github.com/hibernate/hibernate-reactive/pull/197
Thanks a lot
